src/CodeGen_D3D12Compute_Dev.cpp

```diff
@@ -0,0 +1,1154 @@
+#include <sstream>
+#include <algorithm>
+#include <iomanip>
+
+#include "CodeGen_D3D12Compute_Dev.h"
+#include "CodeGen_Internal.h"
+#include "Debug.h"
+#include "IROperator.h"
+
+#define DEBUG_TYPES (1)
+
+namespace Halide {
+namespace Internal {
+
+using std::ostringstream;
+using std::string;
+using std::vector;
+using std::sort;
+
+static ostringstream nil;
+
+CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_Dev(Target t) :
+    d3d12compute_c(src_stream, t) {
+}
+
+string CodeGen_D3D12Compute_Dev::CodeGen_D3D12Compute_C::print_type_maybe_storage(Type type, bool storage, AppendSpaceIfNeeded space) {
+    ostringstream oss;
+
+    // Storage uses packed vector types.
+    if (storage && type.lanes() != 1)
+    {
+        // TODO(marcos): HLSL has 'packoffset', which only applies to constant buffer fields (constant registers (c))
+    }
+    if (type.is_float())
+    {
+        switch (type.bits())
+        {
+            case 16:
+                // 16-bit floating point value. This data type is provided only for language compatibility.
+                // Direct3D 10 shader targets map all half data types to float data types.
+                // A half data type cannot be used on a uniform global variable (use the /Gec flag if this functionality is desired).
+                oss << "half";
+                break;
+            case 32:
+                oss << "float";
+                break;
+            case 64:
+                // 64-bit floating point value. You cannot use double precision values as inputs and outputs for a stream.
+                // To pass double precision values between shaders, declare each double as a pair of uint data types.
+                // Then, use the asdouble function to pack each double into the pair of uints and the asuint function to unpack the pair of uints back into the double.
```

Comment:
Is `asdouble()` exposed to the user?

Reply:
It is, but it is super finicky... You need to encode your double as two unsigned integers.
For now, I'll go ahead and disable `double` types.
We can revisit it if we decide to move to SM 6.x.